File: sonnet-java/src/eval/Evaluator.java
package eval;

import java.util.*;
import java.util.stream.*;
import syntax.AST;
import util.Zipper;

public class Evaluator {
	private final AST.ListNode unit = new AST.ListNode(
			List.of(new AST.IdentNode("quote"), new AST.ListNode(List.of())));

	public AST eval(AST node, Frame frame) {
		AST ast = switch (node) {
			case AST.IntNode n -> n;
			case AST.FloatNode n -> n;
			case AST.StringNode n -> n;
			case AST.IdentNode id -> {
				AST val = frame.lookup(id.name());
				if (val != null)
					yield val;
				throw new EvaluationError("Undefined variable: " + id.name());
			}
			case AST.ListNode list -> {
				List<AST> elements = list.elements();
				if (elements.isEmpty()) {
					yield list; // () -> () symbolic evaluation
				}

				if (isFunction(list)) {
					yield list;
				} else if (isQuote(list)) {
					yield elements.get(1);
				} else if (isConditional(list)) {
					yield evalConditional(elements.get(1), elements.get(2), elements.get(3), frame);
				} else if (isSet(list)) {
					yield evalSet(elements.get(1), elements.get(2), frame);
				}

				// Application

				AST first = elements.get(0);
				List<AST> arguments = elements.subList(1, elements.size());

				if (isBuiltIn(first)) {
					yield evaluateBuiltIn(((AST.IdentNode) first).name(), arguments, frame);
				}

				AST callee = eval(first, frame);

				if (isFunction(callee)) {
					yield evalFunctionApp((AST.ListNode) callee, arguments, frame);
				}

				throw new EvaluationError("First element of the list is not a function or builtin operator: " + callee);
			}
		};
		System.out.println(frame);
		return ast;
	}

	private boolean isFunction(AST f) {
		if (f instanceof AST.ListNode funNode) {
			List<AST> funElements = funNode.elements();
			if (funElements.size() != 3)
				return false;
			if (!(funElements.get(0) instanceof AST.IdentNode id && id.name().equals("fun")))
				return false;
			return isFormals(funElements.get(1));
		}
		return false;
	}

	private boolean isQuote(AST q) {
		if (q instanceof AST.ListNode quoteNode) {
			List<AST> quoteElements = quoteNode.elements();
			if (quoteElements.size() != 2)
				return false;
			if (!(quoteElements.get(0) instanceof AST.IdentNode id && id.name().equals("quote")))
				return false;
			return true;
		}
		return false;
	}

	private boolean isConditional(AST c) {
		if (c instanceof AST.ListNode cNode) {
			List<AST> cElements = cNode.elements();
			if (cElements.size() != 4)
				return false;
			if (!(cElements.get(0) instanceof AST.IdentNode id && id.name().equals("if")))
				return false;
			return true;
		}
		return false;
	}

	private boolean isSet(AST c) {
		if (c instanceof AST.ListNode cNode) {
			List<AST> cElements = cNode.elements();
			if (cElements.size() != 3)
				return false;
			if (!(cElements.get(0) instanceof AST.IdentNode id && id.name().equals("set")))
				return false;
			return true;
		}
		return false;
	}

	private AST evalConditional(AST cond, AST then, AST els, Frame frame) {
		AST c = eval(cond, frame);
		if (c.equals(new AST.IdentNode("true")))
			return eval(then, frame);
		if (c.equals(new AST.IdentNode("false")))
			return eval(els, frame);
		throw new EvaluationError("Conditionals condition must be to either 'true or 'false.");
	}

	private AST evalSet(AST target, AST expr, Frame frame) {
		if (!(target instanceof AST.IdentNode id)) {
			throw new EvaluationError("set assignment target must be a valid identifier symbol.");
		}
		AST calculatedValue = eval(expr, frame);
		frame.assign(id.name(), calculatedValue);
		return unit;
	}

	private AST evalFunctionApp(AST.ListNode callee, List<AST> rawArgs, Frame frame) {
		List<AST> args = rawArgs.stream().map(arg -> eval(arg, frame)).toList();

		Frame bodyFrame = new Frame(frame);

		AST paramNode = callee.elements().get(1);
		List<AST> paramsRaw = (paramNode instanceof AST.ListNode pList) ? pList.elements() : List.of(paramNode);
		List<String> params = paramsRaw.stream().map(e -> {
			if (e instanceof AST.IdentNode id) {
				return id.name();
			}
			{
				throw new EvaluationError("Formals must be a list of identifiers or a singular identifer.");
			}
		}).toList();

		if (params.stream().collect(Collectors.toSet()).size() != params.size()) {
			throw new EvaluationError("It is an error for a variable to appear more than once in formals.");
		}

		AST body = callee.elements().get(2);

		if (params.size() != args.size()) {
			throw new EvaluationError(
					"Arity mismatch: function expects " + params.size() + " arguments, but received " + args.size());
		}

		Zipper.zip(params.stream(), args.stream(), (param, arg) -> {

			bodyFrame.define(param, arg);

		});

		return eval(body, bodyFrame);
	}

	private boolean isFormals(AST f) {
		if (f instanceof AST.IdentNode id)
			return true;
		if (f instanceof AST.ListNode list)
			return list.elements().stream().allMatch(e -> e instanceof AST.IdentNode id);
		return false;
	}

	private boolean isBuiltIn(AST first) {
		if (first instanceof AST.IdentNode id) {
			String name = id.name();
			return name.equals("+") || name.equals("-") || name.equals("*") || name.equals("/") || name.equals("print");
		}
		return false;
	}

	private AST evaluateBuiltIn(String op, List<AST> rawArgs, Frame frame) {
		List<AST> args = rawArgs.stream().map(arg -> eval(arg, frame)).toList();

		return switch (op) {
			case "print" -> {
				StringBuilder sb = new StringBuilder();
				for (int i = 0; i < args.size(); i++) {
					AST arg = args.get(i);
					if (arg instanceof AST.StringNode str) {
						sb.append(str.value());
					} else {
						sb.append(arg.toString());
					}
				}
				System.out.println(sb.toString());
				yield new AST.StringNode(sb.toString());
			}

			case "+", "-", "*", "/" -> {
				boolean isFloat = rawArgs.get(0) instanceof AST.FloatNode node;
				if (isFloat) {
					yield new AST.FloatNode(switch (op) {
						case "+" -> args.stream().reduce(0.0, (acc, arg) -> {
							if (arg instanceof AST.FloatNode f)
								return (acc + f.value());
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a + b);
						case "-" -> args.stream().reduce(null, (acc, arg) -> {
							if (arg instanceof AST.FloatNode f)
								return (acc == null) ? f.value() : acc - f.value();
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a - b);
						case "*" -> args.stream().reduce(0.0, (acc, arg) -> {
							if (arg instanceof AST.FloatNode f)
								return (acc * f.value());
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a * b);
						case "/" -> args.stream().reduce(null, (acc, arg) -> {
							if (arg instanceof AST.FloatNode f)
								return (acc == null) ? f.value() : acc / f.value();
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a / b);
						default ->
							throw new EvaluationError("Impossible state. This would mean an undefined math operation.");
					});
				} else {
					yield new AST.IntNode(switch (op) {
						case "+" -> args.stream().reduce(0, (acc, arg) -> {
							if (arg instanceof AST.IntNode i)
								return (acc + i.value());
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a + b);
						case "-" -> args.stream().reduce(null, (acc, arg) -> {
							if (arg instanceof AST.IntNode i)
								return (acc == null) ? i.value() : acc / i.value();
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a - b);
						case "*" -> args.stream().reduce(0, (acc, arg) -> {
							if (arg instanceof AST.IntNode i)
								return (acc * i.value());
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a * b);
						case "/" -> args.stream().reduce(null, (acc, arg) -> {
							if (arg instanceof AST.IntNode i)
								return (acc == null) ? i.value() : acc / i.value();
							throw new EvaluationError("Mixed  number types in builtin math operation '" + op + "'.");
						}, (a, b) -> a / b);
						default ->
							throw new EvaluationError("Impossible state. This would mean an undefined math operation.");
					});
				}
			}
			default -> throw new EvaluationError("Unknown built-in operator: " + op);
		};
	}

	private double getNumericValue(AST node) {
		if (node instanceof AST.IntNode intNode)
			return intNode.value();
		if (node instanceof AST.FloatNode floatNode)
			return floatNode.value();
		throw new EvaluationError("Expected a numeric node value.");
	}
}

File: sonnet-scala/src/eval/Evaluator.scala
package eval

import syntax.Node
import app.CompilerContext
import syntax.Node.IdentNode
import pprint.pprintln

class Evaluator(ctx: CompilerContext) {

    private var expansionPhase = false

    private val unit: Node.ListNode = Node.ListNode(
      List(Node.IdentNode("quote"), Node.ListNode(List.empty))
    )

    def evaluate(node: Node): Node = {
        // expansionPhase = true;
        eval(node, Frame())
    }

    def eval(node: Node, frame: Frame): Node = {
        val result = node match {
            case n: Node.IntNode    => n
            case n: Node.DoubleNode => n
            case n: Node.StringNode => n

            case Node.IdentNode(name) => {
                frame.lookup(name) match {
                    case Some(value) => value
                    case None        => throw new EvaluationError(s"Undefined variable: $name")
                }
            }

            case list @ Node.ListNode(elements) => {
                val splicedElements = elements.flatMap {
                    case Node.ListNode(Node.IdentNode("splice") :: target :: Nil) =>
                        eval(target, frame) match {
                            case Node.ListNode(innerElements) => innerElements
                            case other                        => throw EvaluationError("Splice can only act upon lists.")
                        }
                    case ordinaryNode =>
                        List(ordinaryNode)
                }

                splicedElements match {
                    case Nil => list // () -> () symbolic evaluation

                    case Node.IdentNode("quote") :: value :: Nil => value

                    case Node.IdentNode("if") :: cond :: thenBr :: elseBr :: Nil =>
                        evalConditional(cond, thenBr, elseBr, frame)

                    case Node.IdentNode("set") :: target :: expr :: Nil =>
                        evalSet(target, expr, frame)
                    case Node.IdentNode("val") :: target :: expr :: Nil =>
                        evalDefine(target, expr, frame, false)
                    case Node.IdentNode("var") :: target :: expr :: Nil =>
                        evalDefine(target, expr, frame, true)

                    case Node.IdentNode("fun") :: formals :: body if isFormals(formals, frame) => list

                    case Node.IdentNode("do") :: body => {
                        val bodyScope = Frame(frame)
                        body.foldLeft[Node](unit)((_, e) => eval(e, bodyScope))
                    }

                    case (n @ Node.IdentNode(op)) :: arguments if isBuiltIn(n) =>
                        evaluateBuiltIn(op, arguments, frame)

                    case first :: arguments => {
                        first match {
                            case IdentNode(name) if isSpecialForm(name) => throw new EvaluationError("Wrong usage of special form.")
                            case _                                      => ()
                        }

                        eval(first, frame) match {
                            case callee: Node.ListNode if callee.elements.headOption.contains(Node.IdentNode("fun")) =>
                                evalFunctionApp(callee, arguments, frame)
                            // case callee: Node.ListNode if callee.elements.headOption.contains(Node.IdentNode("syn")) =>
                            // evalSyntaxApp(callee, arguments, frame)
                            case callee =>
                                throw new EvaluationError(s"Not a function or builtin operator: $callee")
                        }
                    }
                }
            }
        }
        if (ctx.printStackFrames()) println(frame)
        result
    }

    private def isSpecialForm(n: String): Boolean = {
        List("fun", "quote", "let-syn", "if", "set", "do").contains(n)
    }

    private def evalConditional(cond: Node, thenBranch: Node, elseBranch: Node, frame: Frame): Node = {
        eval(cond, frame) match {
            case Node.IdentNode("true")  => eval(thenBranch, frame)
            case Node.IdentNode("false") => eval(elseBranch, frame)
            case _                       => throw new EvaluationError("Conditional condition must evaluate to either 'true or 'false.")
        }
    }

    private def evalSet(target: Node, expr: Node, frame: Frame): Node = {
        target match {
            case Node.IdentNode(name) => {
                val calculatedValue = eval(expr, frame)
                frame.assign(name, calculatedValue)
                unit
            }
            case _ => throw new EvaluationError("set assignment target must be a valid identifier symbol.")
        }
    }

    private def evalDefine(target: Node, expr: Node, frame: Frame, mutable: Boolean): Node = {
        target match {
            case Node.IdentNode(name) => {
                val calculatedValue = eval(expr, frame)
                frame.define(name, calculatedValue, mutable)
                unit
            }
            case _ => throw new EvaluationError("Definition target must be a valid identifier symbol.")
        }
    }

    private def evalFunctionApp(callee: Node.ListNode, rawArgs: List[Node], frame: Frame): Node = {
        val args      = rawArgs.map(arg => eval(arg, frame))
        val bodyFrame = new Frame()

        val paramNode = callee.elements(1)
        val paramsRaw = paramNode match {
            case Node.ListNode(pList) => pList
            case other                => List(other)
        }

        val params = paramsRaw.map {
            case Node.IdentNode(name) => name
            case _                    => throw new EvaluationError("Formals must be a list of identifiers or a singular identifier.")
        }

        if (params.distinct.size != params.size) {
            throw new EvaluationError("It is an error for a variable to appear more than once in formals.")
        }

        val callee_elems = callee.elements

        val body = callee_elems.slice(2, callee_elems.size)

        if (params.size != args.size) {
            throw new EvaluationError(s"Arity mismatch: function expects ${params.size} arguments, but received ${args.size}")
        }

        params.zip(args).foreach { case (param, arg) =>
            bodyFrame.define(param, arg, false)
        }

        body.foldLeft[Node](unit)((_, e) => eval(e, bodyFrame))
    }

    private def evalSyntaxApp(callee: Node.ListNode, args: List[Node], frame: Frame): Node = {
        val bodyFrame = new Frame(frame)

        val paramNode = callee.elements(1)
        val paramsRaw = paramNode match {
            case Node.ListNode(pList) => pList
            case other                => List(other)
        }

        val params = paramsRaw.map {
            case Node.IdentNode(name) => name
            case _                    => throw new EvaluationError("Formals must be a list of identifiers or a singular identifier.")
        }

        if (params.distinct.size != params.size) {
            throw new EvaluationError("It is an error for a variable to appear more than once in formals.")
        }

        val body = callee.elements(2)

        if (params.size != args.size) {
            throw new EvaluationError(s"Arity mismatch: function expects ${params.size} arguments, but received ${args.size}")
        }

        params.zip(args).foreach { case (param, arg) =>
            bodyFrame.define(param, arg, false)
        }

        val expandedAst = substitute(body, bodyFrame)

        eval(expandedAst, frame)
    }

    private def substitute(node: Node, macroFrame: Frame): Node = {
        node match {
            case Node.IdentNode(name) =>
                macroFrame.lookup(name).getOrElse(node)

            case Node.ListNode(elements) =>
                Node.ListNode(elements.map(e => substitute(e, macroFrame)))

            case other =>
                other
        }
    }

    private def isFormals(node: Node, frame: Frame): Boolean = {
        node match {
            case Node.IdentNode(name) if !frame.contains(name) => true
            case Node.ListNode(elements) =>
                elements.forall(element =>
                    element match {
                        case Node.IdentNode(name) => !frame.contains(name)
                        case _                    => false
                    }
                )
            case _ => false
        }
    }

    private def isBuiltIn(node: Node): Boolean = {
        node match {
            case Node.IdentNode(name) => {
                name == "+" || name == "-" || name == "*" || name == "/" || name == "print"
            }
            case _ => false
        }
    }

    private def evaluateBuiltIn(op: String, rawArgs: List[Node], frame: Frame): Node = {
        val args = rawArgs.map(arg => eval(arg, frame))

        op match {
            case "print" => {
                val sb = new StringBuilder()
                args.foreach {
                    case Node.StringNode(value) => sb.append(value)
                    case other                  => sb.append(other.toString)
                }
                println(sb.toString())
                unit
            }

            case "+" | "-" | "*" | "/" => {
                val isDouble = args.head.isInstanceOf[Node.DoubleNode]

                if (isDouble) {
                    val doubleArgs = args.map {
                        case Node.DoubleNode(v) => v
                        case _                  => throw new EvaluationError(s"Mixed number types in builtin math operation $op.")
                    }

                    val calculated = op match {
                        case "+" => doubleArgs.sum
                        case "*" => doubleArgs.product
                        case "-" => doubleArgs.reduceLeft(_ - _)
                        case "/" => doubleArgs.reduceLeft(_ / _)
                        case _   => throw new EvaluationError("Undefined math operation.")
                    }
                    Node.DoubleNode(calculated)
                } else {
                    val intArgs = args.map {
                        case Node.IntNode(v) => v
                        case _               => throw new EvaluationError(s"Mixed number types in builtin math operation $op.")
                    }

                    val calculated = op match {
                        case "+" => intArgs.sum
                        case "*" => intArgs.product
                        case "-" => intArgs.reduceLeft(_ - _)
                        case "/" => intArgs.reduceLeft(_ / _)
                        case _   => throw new EvaluationError("Undefined math operation.")
                    }
                    Node.IntNode(calculated)
                }
            }
            case _ => throw new EvaluationError(s"Unknown built-in operator: $op")
        }
    }
}
